### ISSUE TITLE
Discussion: bringing back @helper for views

### ISSUE THREAD

**Issue by NickCraver** [aspnet/AspNetCore#5110]
In porting some existing applications (from ASP.NET MVC 5), we're seeing immense amounts of pain with views that use `@helper` today. `@helper`, from the user view point, is a very simple way to use Razor templating to re-use some code. Perhaps to render a date a certain way, or a chart, or a graph, or whatever. As an example, Stack Overflow's codebase has 458 `@helper`s today.

Some history:
- [`@helper` was removed in #281](https://github.com/aspnet/Razor/issues/281) after findings in https://github.com/aspnet/Mvc/issues/1130
- Parser errors were added in https://github.com/aspnet/Razor/issues/472,
- A lot of user discussion in https://github.com/aspnet/Razor/issues/715, ultimate closed as "We have no plans to do this"

But `@helper` still lacks an equally (or even close) usable replacement. The alternatives I'm aware of today are:
- [Tag Helpers](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/intro?view=aspnetcore-2.1) which involve writing C# per helper
- [View Components](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/view-components?view=aspnetcore-2.1) which involve writing C# and a view per helper
- Using [a `dynamic` workaround](https://github.com/aspnet/Razor/issues/715#issuecomment-198456302) which isn't type safe and I'm not sure on the performance implications of
- Build the string yourself, and hope you encode everything correctly

There are 2 ways `@helper` worked in ASP.NET MVC 5: globally from App_Data and in individual views. **I am only advocating for support of the latter**. IMO, Tag Helpers or view components (pick your flavor!) are excellent replacements for the global use case. That makes total sense and having another view file is also reasonable.

Let's say we have a view today with 5 `@helper` sections in ASP.NET MVC 5. In ASP.NET Core, it's a regression. We have a few options:
- Copy the code everywhere
- Write a function rendering HTML strings (basically no Razor or safety for the helper)
- Have a child view per helper, going from 1 view to 6, and 1 file to 6
- Create a tag helper for each (5 classes...and again mostly removing HTML safety, or suffering overhead)
- Use the `dynamic` approach, eat the performance, and hope it doesn't go boom at runtime

Note: I am not sure about the performance differences on these. That needs to be tested. I'll try and get benchmarks going when time allows. I have a hard time imagining any of these would be as fast as what `@helper` used to do, though. But I take nothing for a given here...we need to benchmark, even if `@helper` isn't a candidate yet.

A lot of users have chimed in on closed issues above. A lot of migration pain is being felt. Probably much more than was envisioned when `@helper` was removed. This is one of the largest issues I see in our migrations and it's something I'm really feeling right now. Is there a possibility we bring `@helper` or equivalent functionality (re-using some Razor templating inside the same view) back?

cc @DamianEdwards @rynowak

**Comment by NickCraver** [comment]
Noting here: we didn't get pinged on this milestone change (I guess GitHub doesn't do that). Stack Overflow's ASP.NET Core migration is still blocked by this. The delay to preview 3 is very frustrating as it hoses a lot of downstream plans indefinitely.

When will this land?

**Comment by danroth27** [comment]
Hi @NickCraver,

Unfortunately this didn't land for Preview 2. We've been redoing our build infrastructure and unfortunately that took more time than expected.

Since this feature will require VS work to enable the corresponding Razor tooling, we need to align this with work with a corresponding VS release. It's too late at this point to add this to the initial release of VS2019, so our new plan is to get this into the first preview of the first update to VS2019, which should align with .NET Core 3.0 Preview 4. I've update the target milestone accordingly.

Sorry about the inconvenience! Let us know if you have any further questions or concerns.

**Comment by KLuuKer** [comment]
We also used this allot in exactly the same way
having some local function to build nested html for some big tree's
and other stuff that was just repeated a bunch of times
great to see this feature come back!

**Comment by NTaylorMullen** [comment]
Added the ability to write markup in `@functions {...}` and local functions (`@{ void Foo() { ... } }`
![image](https://user-images.githubusercontent.com/2008729/54706661-a9a6b680-4afc-11e9-86c2-bbd7c047e54a.png)

https://github.com/aspnet/AspNetCore-Tooling/commit/5ffd84e56dd3d21b48d55feea5bc1493d0293fd1
